### PR TITLE
Fix for loading imageset with '%' char in filepath

### LINKED ISSUE #214: ImageSetFactory.make_sequence(): "%" in template breaks format string
Reproduced with:

```
$ ln -s $(dials.data get -q centroid_test_data) test%
$ dials.import test%/centroid_000*.cbf
DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
DIALS 3.dev.174-gd1f0e33e8
Traceback (most recent call last):
  File "/Users/rjgildea/software/cctbx_py3/build/../../cctbx/modules/dials/command_line/dials_import.py", line 923, in <module>
    script.run()
  File "/Users/rjgildea/software/cctbx_py3/build/../../cctbx/modules/dials/command_line/dials_import.py", line 754, in run
    params, options = self.parser.parse_args(args=args, show_diff_phil=False)
  File "/Users/rjgildea/software/cctbx/modules/dials/util/options.py", line 841, in parse_args
    quick_parse=quick_parse,
  File "/Users/rjgildea/software/cctbx/modules/dials/util/options.py", line 554, in parse_args
    load_models=load_models,
  File "/Users/rjgildea/software/cctbx/modules/dials/util/options.py", line 212, in __init__
    load_models,
  File "/Users/rjgildea/software/cctbx/modules/dials/util/options.py", line 284, in try_read_experiments_from_images
    load_models=load_models,
  File "/Users/rjgildea/software/cctbx/modules/dxtbx/model/experiment_list.py", line 536, in from_filenames
    format_kwargs=format_kwargs,
  File "/Users/rjgildea/software/cctbx/modules/dxtbx/datablock.py", line 1501, in from_filenames
    format_kwargs=format_kwargs,
  File "/Users/rjgildea/software/cctbx/modules/dxtbx/datablock.py", line 974, in __init__
    imagesets = list(imagesets)
  File "/Users/rjgildea/software/cctbx/modules/dxtbx/datablock.py", line 841, in _convert_to_imagesets
    yield _create_imagesequence(setgroup[0], format_class, format_kwargs)
  File "/Users/rjgildea/software/cctbx/modules/dxtbx/datablock.py", line 799, in _create_imagesequence
    format_kwargs=format_kwargs,
  File "/Users/rjgildea/software/cctbx/modules/dxtbx/imageset.py", line 563, in make_sequence
    filenames = [template_format % index for index in indices]
  File "/Users/rjgildea/software/cctbx/modules/dxtbx/imageset.py", line 563, in <listcomp>
    filenames = [template_format % index for index in indices]
ValueError: Please report this error to dials-support@lists.sourceforge.net: unsupported format character '/' (0x2f) at index 25
```

Failing line of code:
https://github.com/cctbx/dxtbx/blob/3fad1e5a6283191cb94c53e3a07ac3a28c1a6cc8/imageset.py#L563

### PR DESCRIPTION
**Pull request by rjgildea**
Fixes #214

### CODE REVIEW

**Comment by ndevenish** [comment]
I've merged our two approaches so that the abstraction-refactor is on top of the move, this seemed reasonable.

I'm also picking the defensive-fix from https://github.com/dials/dxtbx/commit/b18eb65350083e394aad8f853d934c7ba97a667a for the release branch, since this includes refactoring now.

File: tests/test_imageset.py [hunk 1]
```diff
@@ -493,6 +493,30 @@ def test_imagesetfactory(centroid_files, dials_data):
...
+    images = [
+        dials_data("centroid_test_data").join(f"centroid_{i:04}.cbf")
+        for i in range(1, 10)
+    ]
+    directory = tmp_path / "test%"
+    directory.mkdir()
+    for image in images:
+        (directory / image.basename).symlink_to(image)
+    template = str(directory / "centroid_####.cbf")
+    sequence = ImageSetFactory.make_sequence(template, list(range(1, 10)))
```
Comment (Anthchirp) [inline tests/test_imageset.py:506]:
```suggestion
    sequence = ImageSetFactory.make_sequence(template, range(1, 10))
```
Code suggest this doesn't need to be list

File: imageset.py [hunk 2]
```diff
@@ -19,6 +19,11 @@
 
 ext = boost_adaptbx.boost.python.import_ext("dxtbx_ext")
 
+try:
+    from typing import Iterable, List
+except ImportError:
+    pass
```
Comment (Anthchirp) [inline imageset.py:25]:
```suggestion
from typing import Iterable, List
```

I believe typing is always present

File: imageset.py [hunk 3]
```diff
@@ -33,6 +38,20 @@
 )
 
 
+def _expand_template(template, indices):
+    # type: (str, Iterable[int]) -> List[str]
```
Comment (Anthchirp) [inline imageset.py:42]:
```suggestion
def _expand_template(template: str, indices: Iterable[int]) -> List[str]:
```
typing syntax is probably more helpful (untested, not sure I'm doing this right)

**Pull request merged**